compiler/rustc_middle/src/ty/util.rs

```diff
@@ -1041,6 +1041,42 @@ pub fn needs_drop_components<'tcx>(
     }
 }
 
+pub fn trivial_const_drop<'tcx>(ty: Ty<'tcx>) -> bool {
+    match *ty.kind() {
+        ty::Bool
+        | ty::Char
+        | ty::Int(_)
+        | ty::Uint(_)
+        | ty::Float(_)
+        | ty::Infer(ty::IntVar(_))
+        | ty::Infer(ty::FloatVar(_))
+        | ty::Str
+        | ty::RawPtr(_)
+        | ty::Ref(..)
+        | ty::FnDef(..)
+        | ty::FnPtr(_)
+        | ty::Never => true,
+
+        ty::Opaque(..)
+        | ty::Dynamic(..)
+        | ty::Error(_)
+        | ty::Bound(..)
+        | ty::Param(_)
+        | ty::Placeholder(_)
+        | ty::Foreign(_)
```

Comment:
I don't think `ty::Foreign` should return false because there is nothing non-const-droppy about them. You cannot have custom drop impls for them i.e. the following would fail:

```rust
#![feature(extern_types)]
extern "C" {
    type Foo;
}

impl Drop for Foo {
    fn drop(&mut self) {
        // side effect
    }
}
```

As for "do we ever encounter this type when checking for const Drop?" I'd say not currently, but if we were to introduce const heap it might become possible:

```rust
#![feature(extern_types)]
extern "C" {
    type Foo;
}

fn dropbox(_x: Box<Foo>) {}
```